gnd/src/main/java/com/google/android/gnd/ui/home/mapcontainer/PolygonDrawingViewModel.java

```diff
@@ -0,0 +1,255 @@
+import com.google.android.gnd.system.auth.AuthenticationManager;
+import com.google.android.gnd.ui.common.AbstractViewModel;
+import com.google.android.gnd.ui.common.SharedViewModel;
+import com.google.common.collect.ImmutableList;
+import io.reactivex.BackpressureStrategy;
+import io.reactivex.Flowable;
+import io.reactivex.Observable;
+import io.reactivex.processors.BehaviorProcessor;
+import io.reactivex.subjects.PublishSubject;
+import io.reactivex.subjects.Subject;
+import java.util.ArrayList;
+import java.util.List;
+import java8.util.Optional;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import timber.log.Timber;
+
+@SharedViewModel
+public class PolygonDrawingViewModel extends AbstractViewModel {
+
+  /** Minimum distance (in metres) between points to be considered as non-overlapping. */
+  public static final int DISTANCE_THRESHOLD = 10;
+
+  @Hot private final Subject<Nil> defaultMapMode = PublishSubject.create();
+
+  @Hot private final Subject<Nil> drawingCompleted = PublishSubject.create();
+
+  private final MutableLiveData<Integer> completeButtonVisible = new MutableLiveData<>(INVISIBLE);
+  /** Polyline drawn by the user but not yet saved as polygon. */
+  @Hot
+  private final MutableLiveData<PolygonFeature> drawnPolylineVertices = new MutableLiveData<>();
+
+  @Hot(replays = true)
+  private final MutableLiveData<Boolean> locationLockEnabled = new MutableLiveData<>();
+
+  private final LiveData<Integer> iconTint;
+  @Hot private final Subject<Boolean> locationLockChangeRequests = PublishSubject.create();
+  private final LocationManager locationManager;
+  private final LiveData<BooleanOrError> locationLockState;
+  private final List<Point> vertices = new ArrayList<>();
+  /** The currently selected layer and project for the polygon drawing. */
+  private final BehaviorProcessor<Optional<Layer>> selectedLayer =
+      BehaviorProcessor.createDefault(Optional.empty());
+
+  private final BehaviorProcessor<Optional<Project>> selectedProject =
+      BehaviorProcessor.createDefault(Optional.empty());
+
+  private final OfflineUuidGenerator uuidGenerator;
+  private final AuthenticationManager authManager;
+  @Nullable private Point cameraTarget;
+
+  @Inject
+  PolygonDrawingViewModel(
+      LocationManager locationManager,
+      AuthenticationManager authManager,
+      OfflineUuidGenerator uuidGenerator) {
+    this.locationManager = locationManager;
+    this.authManager = authManager;
+    this.uuidGenerator = uuidGenerator;
+    Flowable<BooleanOrError> locationLockStateFlowable = createLocationLockStateFlowable().share();
+    this.locationLockState =
+        LiveDataReactiveStreams.fromPublisher(
+            locationLockStateFlowable.startWith(BooleanOrError.falseValue()));
+    this.iconTint =
+        LiveDataReactiveStreams.fromPublisher(
+            locationLockStateFlowable
+                .map(locked -> locked.isTrue() ? R.color.colorMapBlue : R.color.colorGrey800)
```

Comment:
This expression can be moved to the layout xml.

Reply:
binding does not work with polygonview model.